### PR TITLE
Various Improvements

### PR DESCRIPTION
**Pull request by happymimimix**
UI Optimizations
Note counter
A scoring system similar to Phigros. 
Polyphony
Unused variables and components removal
Useless files removal
CC151 Pitch bend sensitivity support
Show custom keys to 128 keys range
More debug information
Operates ffmpeg automatically
Support various visual effects via cheat engine, e.g.
Move and zoom piano keyboard at specific time in the music
Hide keyboard
Change playback speed
Change note falling speed
Video output compatible

Lua script template: 

TME = "PianoFromAboveVIZ.exe+1E12E0"
PBS = "PianoFromAboveVIZ.exe+1E1178"
NTS = "PianoFromAboveVIZ.exe+1E1180"
PBV = "PianoFromAboveVIZ.exe+1E1188"
ZOM = "PianoFromAboveVIZ.exe+1E11A8"
MVX = "PianoFromAboveVIZ.exe+1E11A0"
MVY = "PianoFromAboveVIZ.exe+1E11A4"
KBD = "PianoFromAboveVIZ.exe+1E119A"
function RTME()
return readDouble(TME)
end
function RPBS()
return readDouble(PBS)
end
function RNTS()
return readDouble(NTS)
end
function RPBV()
return readDouble(PBV)
end
function RZOM()
return readFloat(ZOM)
end
function RMVX()
return readFloat(MVX)
end
function RMVY()
return readFloat(MVY)
end
function RKBD()
return readByte(KBD)
end
function WTME(VAL)
writeDouble(TME,VAL)
end
function WPBS(VAL)
writeDouble(PBS,VAL)
end
function WNTS(VAL)
writeDouble(NTS,VAL)
end
function WPBV(VAL)
writeDouble(PBV,VAL)
end
function WZOM(VAL)
writeFloat(ZOM,VAL)
end
function WMVX(VAL)
writeFloat(MVX,VAL)
end
function WMVY(VAL)
writeFloat(MVY,VAL)
end
function KBDO()
writeByte(KBD,0)
end
function KBDI()
writeByte(KBD,1)
end
function KBDT()
if (RKBD() == 1)then
writeByte(KBD,0)
end
if (RKBD() == 0)then
writeByte(KBD,1)
end
end
function WIT(TRM)
while (RTME() < TRM) do
end
end

--Start scripting from here

Color events comming soon!

### CODE REVIEW

File: LICENSE.WojciechMula-toys
Comment (khang06) [inline LICENSE.WojciechMula-toys:1]: This file is legally required under that license to be included with this project.

File: PianoFromAbove/Config.cpp [hunk 1]
```diff
@@ -75,7 +75,7 @@ void Config::LoadConfigValues()
     LoadConfigValues( txRoot );
 
     // Custom settings need to be loaded from a separate file, otherwise stock PFA will reset them
-    doc = TiXmlDocument(sPath + "\\pfavizkhang.xml");
+    doc = TiXmlDocument(sPath + "\\Config-VIZ.xml");
```
Comment (khang06) [inline PianoFromAbove/Config.cpp:78]: I don't see a reason to rename these. It'll force everyone to restore all of their configuration changes and it'll prevent settings from syncing with stock PFA.

File: PianoFromAbove/Config.cpp [hunk 2]
```diff
@@ -177,8 +177,8 @@ void VideoSettings::LoadDefaultValues()
 
 void ControlsSettings::LoadDefaultValues()
 {
-    this->dFwdBackSecs = 3.0;
-    this->dSpeedUpPct = 10.0;
+    this->dFwdBackSecs = 5;
+    this->dSpeedUpPct = 10;
```
Comment (khang06) [inline PianoFromAbove/Config.cpp:181]: I don't see why this has to be changed.

File: PianoFromAbove/GameState.cpp [hunk 3]
```diff
@@ -266,15 +276,6 @@ GameState::GameError SplashScreen::MsgProc( HWND hWnd, UINT msg, WPARAM wParam,
...
             break;
-        case WM_KEYDOWN:
-        {
-            switch( wParam )
-            {
-                case VK_SPACE:
-                    cPlayback.TogglePaused( true );
-                    return Success;
-            }
-        }
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:277]: I don't see a reason to remove this.

File: PianoFromAbove/GameState.cpp [hunk 4]
```diff
@@ -660,19 +666,12 @@ void MainScreen::InitState()
...
-        SetWindowTextA(g_hWnd, buf);
-        m_hVideoPipe = CreateNamedPipe(TEXT("\\\\.\\pipe\\pfadump"),
-            PIPE_ACCESS_OUTBOUND,
-            PIPE_TYPE_BYTE | PIPE_WAIT,
-            PIPE_UNLIMITED_INSTANCES,
-            static_cast<DWORD>(width * height * 4 * 120),
-            0,
-            0,
-            nullptr);
+        snprintf(buf, sizeof(buf), "%s&cd \"%s\"&md \"%s\\PFA_Framedump\"&start cmd /k \"ffmpeg -r 60 -f rawvideo -s %dx%d -pixel_format rgb32 -i \"async:\\\\.\\pipe\\pfadump\" -c:v h264_nvenc -qp 0 ^\"%s\\PFA_Framedump\\Output.mkv^\"\"", GetExePath().substr(0, 2).c_str(), GetExePath().c_str(), GetExePath().c_str(), width, height, GetExePath().c_str());
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:671]: This kind of thing really has to be configurable. Like, what if someone wasn't on a NVIDIA GPU?

Also, you should use the proper Windows APIs for things like creating directories.

File: PianoFromAbove/GameState.cpp [hunk 5]
```diff
@@ -843,77 +845,31 @@ GameState::GameError MainScreen::MsgProc( HWND hWnd, UINT msg, WPARAM wParam, LP
...
-                    return Success;
-                case VK_RIGHT:
-                    JumpTo(static_cast<long long>(m_llStartTime + cControls.dFwdBackSecs * 1000000));
-                    return Success;
-                case 'M':
-                    cPlayback.ToggleMute( true );
-                    return Success;
-            }
-            break;
-        }
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:916]: Why would you remove all of these hotkeys? People do actually use some of them.

File: PianoFromAbove/GameState.cpp [hunk 6]
```diff
@@ -988,18 +945,18 @@ GameState::GameError MainScreen::MsgProc( HWND hWnd, UINT msg, WPARAM wParam, LP
...
-    static const VideoSettings &cVideo = config.GetVideoSettings();
-    static const VizSettings &cViz = config.GetVizSettings();
-    const MIDI::MIDIInfo &mInfo = m_MIDI.GetInfo();
+    static Config& config = Config::GetConfig();
+    static PlaybackSettings& cPlayback = config.GetPlaybackSettings();
+    static const ViewSettings& cView = config.GetViewSettings();
+    static const VisualSettings& cVisual = config.GetVisualSettings();
+    static const VideoSettings& cVideo = config.GetVideoSettings();
+    static const VizSettings& cViz = config.GetVizSettings();
+    const MIDI::MIDIInfo& mInfo = m_MIDI.GetInfo();
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:959]: I know you were working without git for a while, but if you want to PR something to someone's project, please try to keep formatting changes out of it (or at least separate it into a separate commit). It makes it a lot harder to pick out actual code changes and leads to unnecessary merge conflicts down the line.

File: PianoFromAbove/GameState.cpp [hunk 7]
```diff
@@ -1105,18 +1064,39 @@ GameState::GameError MainScreen::Logic( void )
...
+                    } else if (pEvent->GetParam1() == 6) {
+                        if (m_RPN_MSB[pEvent->GetChannel()] && m_RPN_LSB[pEvent->GetChannel()]) {
+                            m_pBendsRange[pEvent->GetChannel()] = pEvent->GetParam2();
+                        }
+                        else {
+                            m_RPN_MSB[pEvent->GetChannel()] = false;
+                            m_RPN_LSB[pEvent->GetChannel()] = false;
+                        }
+                    }
+                }
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1089]: This doesn't handle seeking. Also, it might be a good idea to add the RPN event codes to ChannelEventType to reduce the amount of "magic numbers" in this block.

File: PianoFromAbove/GameState.cpp [hunk 8]
```diff
@@ -1105,18 +1064,39 @@ GameState::GameError MainScreen::Logic( void )
...
+                else if (config.m_bPianoOverride && pEvent->GetChannelEventType() == MIDIChannelEvent::ProgramChange) {
+                    pEvent->SetParam1(0);
                 }
                 m_OutDevice.PlayEvent(pEvent->GetEventCode(), pEvent->GetParam1(), pEvent->GetParam2());
             }
             else if (!m_bMute && !m_vTrackSettings[pEvent->GetTrack()].aChannels[pEvent->GetChannel()].bMuted) {
                 m_OutDevice.PlayEvent(pEvent->GetEventCode(), pEvent->GetParam1(),
                     static_cast<int>(pEvent->GetParam2() * dVolumeCorrect + 0.5));
                 notes_played++;
+                m_iNoteCount++;
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1099]: This also doesn't work well because it doesn't handle seeking, and is exactly why I don't have a note counter.

File: PianoFromAbove/GameState.cpp [hunk 9]
```diff
@@ -1145,14 +1125,12 @@ GameState::GameError MainScreen::Logic( void )
...
         
         // Update NPS
-        if (cViz.bNerdStats) {
             for (; !m_dNPSNotes.empty(); m_dNPSNotes.pop_front()) {
                 if (std::get<0>(m_dNPSNotes.front()) >= m_llStartTime - 1000000)
                     break;
             }
             if (notes_played != 0)
                 m_dNPSNotes.push_back(std::make_tuple(m_llStartTime, notes_played));
-        }
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1155]: I don't see a reason to make this default.

File: PianoFromAbove/GameState.cpp [hunk 10]
```diff
@@ -1261,7 +1250,7 @@ void MainScreen::JumpTo(long long llStartTime, bool bUpdateGUI)
         auto itPrev = itMiddle - 1;
         int iFound = 0;
         int iSimultaneous = m_vEvents[itPrev - m_vEvents.begin()]->GetSimultaneous() + 1;
-        for (std::vector<MIDIChannelEvent*>::reverse_iterator it(itMiddle); iFound < iSimultaneous && it != m_vEvents.rend(); ++it)
+        for (std::vector<MIDIChannelEvent*>::reverse_iterator it(itBegin); iFound < iSimultaneous && it != m_vEvents.rend(); ++it)
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1253]: Why? This change breaks seeking by removing every note that's touching the keyboard at that time.

File: PianoFromAbove/GameState.cpp [hunk 11]
```diff
@@ -1288,59 +1277,10 @@ void MainScreen::JumpTo(long long llStartTime, bool bUpdateGUI)
...
-            !aProgram[pEvent->GetChannel()])
-        {
-            aProgram[pEvent->GetChannel()] = true;
-            m_OutDevice.PlayEvent(pEvent->GetEventCode(), pEvent->GetParam1(), pEvent->GetParam2());
-        }
-    }
-
-    // Finally play the controller events. vControl is in reverse time order
-    for (vector< MIDIChannelEvent* >::reverse_iterator it = vControl.rbegin(); it != vControl.rend(); ++it)
-        m_OutDevice.PlayEvent((*it)->GetEventCode(), (*it)->GetParam1(), (*it)->GetParam2());
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1343]: Why did this get removed? This breaks things like program changes being properly updated when seeking.

File: PianoFromAbove/GameState.cpp [hunk 12]
```diff
@@ -1577,7 +1517,6 @@ GameState::GameError MainScreen::Render()
     RenderNotes();
     if ( m_bShowKB )
         RenderKeys();
-    RenderBorder();
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1580]: I don't see a reason to remove this.

File: PianoFromAbove/GameState.cpp [hunk 13]
```diff
@@ -1588,15 +1527,8 @@ GameState::GameError MainScreen::Render()
...
         auto* frame = m_pRenderer->Screenshot();
-
         // Write to pipe
         WriteFile(m_hVideoPipe, frame, static_cast<DWORD>(m_pRenderer->GetBufferWidth() * m_pRenderer->GetBufferHeight() * 4), nullptr, nullptr);
-
-        // Show dump speed on the title bar
-        const std::wstring& name = m_MIDI.GetInfo().sFilename;
-        TCHAR sTitle[1024];
-        _stprintf_s(sTitle, TEXT("%ws (%.1lf%%)"), name.c_str() + (name.find_last_of(L'\\') + 1), (m_dFPS / m_Timer.m_dFramerate) * 100.0);
-        SetWindowText(g_hWnd, sTitle);
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1599]: This was a pretty useful feature. I don't know why you removed it.

File: PianoFromAbove/GameState.cpp [hunk 14]
```diff
@@ -1865,14 +1797,6 @@ void MainScreen::RenderKeys()
...
                 m_pRenderer->DrawRect( fCurX + fKeyGap1, fCurY + fTopCY, m_fWhiteCX - fKeyGap, 2.0f,
                     m_csKBBackground.iDarkRGB, m_csKBBackground.iDarkRGB, m_csKBWhite.iVeryDarkRGB, m_csKBWhite.iVeryDarkRGB );
-
-                if ( i == MIDI::C4 )
-                {
-                    float fMXGap = floor( m_fWhiteCX * 0.25f + 0.5f );
-                    float fMCX = m_fWhiteCX - fMXGap * 2.0f - fKeyGap;
-                    float fMY = max( fCurY + fTopCY - fMCX - 5.0f, fCurY + fSharpCY + 5.0f );
-                    m_pRenderer->DrawRect( fCurX + fKeyGap1 + fMXGap, fMY, fMCX, fCurY + fTopCY - 5.0f - fMY, m_csKBWhite.iDarkRGB );
-                }
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1875]: Again, why was this removed? If it's supposed to be for performance reasons, rendering a single tiny square isn't going to do much to performance.

File: PianoFromAbove/GameState.cpp [hunk 15]
```diff
@@ -1900,8 +1816,8 @@ void MainScreen::RenderKeys()
         }
 
     // Draw the sharps
-    iStartRender = ( m_iStartNote != MIDI::A0 && !MIDI::IsSharp( m_iStartNote ) && m_iStartNote > 0 && MIDI::IsSharp( m_iStartNote - 1 ) ? m_iStartNote - 1 : m_iStartNote );
-    iEndRender = ( m_iEndNote != MIDI::C8 && !MIDI::IsSharp( m_iEndNote ) && m_iEndNote < 127 && MIDI::IsSharp( m_iEndNote + 1 ) ? m_iEndNote + 1 : m_iEndNote );
+    iStartRender = ( m_iStartNote != 0 && !MIDI::IsSharp( m_iStartNote ) && m_iStartNote > 0 && MIDI::IsSharp( m_iStartNote - 1 ) ? m_iStartNote - 1 : m_iStartNote );
+    iEndRender = ( m_iEndNote != 127 && !MIDI::IsSharp( m_iEndNote ) && m_iEndNote < 127 && MIDI::IsSharp( m_iEndNote + 1 ) ? m_iEndNote + 1 : m_iEndNote );
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1820]: I don't get what the point of this change is supposed to be.

File: PianoFromAbove/GameState.cpp [hunk 16]
```diff
@@ -2063,41 +1948,144 @@ void MainScreen::RenderStatusLine(int line, const char* left, const char* format
...
-    auto starttime = ((m_llStartTime >= 0) ? m_llStartTime : -m_llStartTime);
-
-    auto min = starttime / 60000000;
-    auto sec = (starttime % 60000000) / 1000000;
-    auto cs = (starttime % 1000000) / 100000;
-    auto tmin = mInfo.llTotalMicroSecs / 60000000;
-    auto tsec = (mInfo.llTotalMicroSecs % 60000000) / 1000000;
-    auto tcs = (mInfo.llTotalMicroSecs % 1000000) / 100000;
-    auto tempo = 60000000.0 / m_iMicroSecsPerBeat;
-
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:2079]: I think you forgot to merge this change into your local repo.

File: PianoFromAbove/GameState.cpp [hunk 17]
```diff
@@ -2063,41 +1948,144 @@ void MainScreen::RenderStatusLine(int line, const char* left, const char* format
...
-    // Nerd stats
-    if (viz.bNerdStats) {
-        long long nps = 0;
-        for (int i = 0; i < m_dNPSNotes.size(); i++)
-            nps += std::get<1>(m_dNPSNotes[i]);
-
-        RenderStatusLine(cur_line++, "NPS:", "%lld", nps);
-        RenderStatusLine(cur_line++, "Rendered:", "%llu", m_pRenderer->GetRenderedNotesCount());
+    RenderStatusLine(cur_line++,"PianoFromAbove VIZ", "v3.1.0");
+    RenderStatusLine(cur_line++,"Made by: happy_mimimix", "");
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1966]: PFA Viz wasn't made by you, even if you have a couple of your own changes in there.

File: PianoFromAbove/GameState.cpp [hunk 18]
```diff
@@ -2063,41 +1948,144 @@ void MainScreen::RenderStatusLine(int line, const char* left, const char* format
...
+    if (m_llStartTime >= 0) {
+        RenderStatusLine(cur_line++, "Time:", "%lld:%04.1lf / %lld:%04.1lf", m_llStartTime / 60000000, (m_llStartTime % 60000000) / 1000000.0, mInfo.llTotalMicroSecs / 60000000, (mInfo.llTotalMicroSecs % 60000000) / 1000000.0);
+    }
+    else {
+        RenderStatusLine(cur_line++, "Time:", "\t-%lld:%04.1lf / %lld:%04.1lf", -m_llStartTime / 60000000, (-m_llStartTime % 60000000) / 1000000.0, mInfo.llTotalMicroSecs / 60000000, (mInfo.llTotalMicroSecs % 60000000) / 1000000.0);
+    }
+    if (m_bDebug) RenderStatusLine(cur_line++, "Milliseconds:", "%.0lf", m_msTime);
+    RenderStatusLine(cur_line++,"FPS:", "%.2lf", m_dFPS);
+    RenderStatusLine(cur_line++,"Tempo:", "%.4lf", 60000000.0 / m_iMicroSecsPerBeat);
+    if (viz.bPhigros) {
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:1981]: I know what Phigros is, but I don't know what that has to do with PFA.

File: PianoFromAbove/GameState.cpp [hunk 19]
```diff
@@ -2063,41 +1948,144 @@ void MainScreen::RenderStatusLine(int line, const char* left, const char* format
...
+            if (!cPlayback.GetPaused()) {
+                RenderStatusLine(cur_line++,"Deomonstration in progress... ", "");
+            }
+            else {
+                RenderStatusLine(cur_line++, "Game paused... ", "");
+            }
+        }
+        else if (static_cast<float>(floor(static_cast<float>(m_iNoteCount) / static_cast<float>(mInfo.iNoteCount) * 1000000)) > static_cast<float>(1000000)) {
+            RenderStatusLine(cur_line++,"ERROR: Score overflow! ", "");
+        }
```
Comment (khang06) [inline PianoFromAbove/GameState.cpp:2088]: Feel free to keep something like this in your fork, but I don't think this is something I'd merge.

File: PianoFromAbove/GameState.h [hunk 20]
```diff
@@ -20,6 +20,7 @@ using namespace std;
 #include "Renderer.h"
 #include "MIDI.h"
 #include "Misc.h"
+#include "robin_hood.h"
```
Comment (khang06) [inline PianoFromAbove/GameState.h:23]: Why did this get readded? It isn't used anywhere anymore.

File: PianoFromAbove/Images/FunkyStars.mid
Comment (khang06) [inline PianoFromAbove/Images/FunkyStars.mid:1]: I don't want to change the default splash MIDI, especially when its replacement makes the executable size 2x larger.

File: PianoFromAbove/Images/PFA Icon.ico
Comment (khang06) [inline PianoFromAbove/Images/PFA Icon.ico:1]: Why change the program icon to a stolen Synthesia icon?

File: PianoFromAbove/MainProcs.cpp [hunk 21]
```diff
@@ -48,9 +48,9 @@ INT_PTR SetResolutionProc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam) {
         case IDOK: {
             char buf[1024] = {};
             GetWindowTextA(GetDlgItem(hwnd, IDC_WIDTH), buf, sizeof(buf));
-            int width = max(MINWIDTH, min(atoi(buf), 65535));
+            int width = max(0, min(atoi(buf), INFINITY));
```
Comment (khang06) [inline PianoFromAbove/MainProcs.cpp:51]: `INFINITY` is meant to be used as a floating point constant. I feel like what you meant was `max(0, atoi(buf));`.

File: PianoFromAbove/MainProcs.cpp [hunk 22]
```diff
@@ -124,12 +124,10 @@ LRESULT WINAPI WndProc( HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam )
                     cPlayback.SetPlayMode( GameState::Intro, true );
                     cPlayback.SetPlayable( false, true );
                     cPlayback.SetPosition( 0 );
-                    SetWindowText( g_hWnd, L"pfavizkhang-dx12 " __DATE__ );
+                    SetWindowText( g_hWnd, L"PianoFromAbove VIZ | Made by: happy_mimimix | Ver 3.1.0 | Now playing: None");
```
Comment (khang06) [inline PianoFromAbove/MainProcs.cpp:127]: Again, you didn't make PFA Viz.

File: PianoFromAbove/MainProcs.cpp [hunk 23]
```diff
@@ -209,9 +205,6 @@ LRESULT WINAPI WndProc( HWND hWnd, UINT msg, WPARAM wParam, LPARAM lParam )
                 case ID_GAMEERROR:
                     MessageBoxW( hWnd, GameState::Errors[lParam].c_str(), L"Error", MB_OK | MB_ICONEXCLAMATION );
                     return 0;
-                case ID_UPDATE:
-                    ShellExecute(NULL, L"open", L"https://github.com/khang06/PianoFromAbove/releases", NULL, NULL, SW_SHOWNORMAL);
-                    return 0;
```
Comment (khang06) [inline PianoFromAbove/MainProcs.cpp:214]: Why did you remove the update checker?

File: PianoFromAbove/MainProcs.cpp [hunk 24]
```diff
@@ -1110,7 +1103,7 @@ INT_PTR LoadingProc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam) {
     }
     case WM_TIMER: {
         // lots of race conditions possible but hopefully nobody notices them
-        const char* desc = "placeholder";
+        const char* desc = "Loading...";
```
Comment (khang06) [inline PianoFromAbove/MainProcs.cpp:1106]: This was intentional. This message wouldn't show up unless something went *very* wrong with the MIDI loader.

File: PianoFromAbove/PianoFromAbove.cpp [hunk 25]
```diff
@@ -40,76 +37,6 @@ HWND g_hWndGfx = NULL;
...
-        }
-    }
-
-    WinHttpCloseHandle(request);
-    WinHttpCloseHandle(connect);
-    WinHttpCloseHandle(session);
-
-    return 0;
-}
-
```
Comment (khang06) [inline PianoFromAbove/PianoFromAbove.cpp:112]: Why did all of the updater code get completely removed?

File: PianoFromAbove/Renderer.cpp [hunk 26]
```diff
@@ -88,10 +88,6 @@ std::tuple<HRESULT, const char*> D3D12Renderer::Init(HWND hWnd, bool bLimitFPS)
             //return std::make_tuple(res, "D3D12CreateDevice");
         break;
     }
-    if (m_pDevice == nullptr) {
-        MessageBoxW(NULL, L"Couldn't find a suitable D3D12 device.", L"DirectX Error", MB_ICONERROR);
-        exit(1);
-    }
```
Comment (khang06) [inline PianoFromAbove/Renderer.cpp:94]: Why would you remove this check? Instead of giving the user a helpful message, removing this will cause it to crash without any error message.

File: PianoFromAbove/Renderer.h [hunk 27]
```diff
@@ -123,8 +123,8 @@ class D3D12Renderer {
     bool UploadBackgroundBitmap();
 
     static constexpr unsigned FrameCount = 2;
-    static constexpr unsigned RectsPerPass = 10000; // Relatively low limit, but not many rects are supposed to be rendered anyway
-    static constexpr unsigned NotesPerPass = 5000000;
+    static constexpr unsigned RectsPerPass = 1048576;
+    static constexpr unsigned NotesPerPass = 1048576;
```
Comment (khang06) [inline PianoFromAbove/Renderer.h:127]: Why were these changed?

File: PianoFromAbove/robin_hood.h
Comment (khang06) [inline PianoFromAbove/robin_hood.h:1]: This file isn't being used anymore.

File: README.md
Comment (khang06) [inline README.md:1]: Why would you delete the entire README?

**Comment by khang06** [comment]
I'm sorry. I want to put this in the nicest way possible, but I really can't merge this PR. Instead of just closing it without saying much, I tried to give you some feedback that you can hopefully find valuable.

File: README.md
Comment (rin-w) [reply README.md:1]: > Why would you delete the entire README?

this optimizes compile time. he could've also removed all the new lines and useless spaces in the code. but still LGTM
